File: PHP/PenPaper/Pen.php
<?php
    require_once 'Paper.php';
    require_once 'Exceptions/OutOfInk.php';

    class Pen {
        private $inkAmount, $inkCapacity;
        
        private function validate($value) {
            if ( is_int($value) && $value > 0 ) {
                return $value;
            }
            throw new InvalidParam("Invalid param '$value'.");
        }
        
        public function __construct($inkCapacity = 4096) {
            $this->inkAmount = $this->validate($inkCapacity);
            $this->inkCapacity = $inkCapacity;
        }
        
        public function __get($name) {
            if ( property_exists($this, $name) ) {
                return $this->$name;
            }
            throw new InvalidAttribute("Attribute error: attribute '$name' not found.");
        }
        
        public function write(Paper $paper, $msg) {
            if ( $this->inkAmount == 0 ) {
                throw new OutOfInk("Please, refill your pen!");
            }
            if ( strlen($msg) > $this->inkAmount ) {
                $paper->addContent(substr($msg, 0, $this->inkAmount));
                $this->inkAmount = 0;
                return;
            }
            $paper->addContent($msg);
            $this->inkAmount -= strlen($msg);
        }
        
        public function refill() {
            $this->inkAmount = $this->inkCapacity;
        }
    }
?>

File: W5/Pen&Paper/Pen.py
from Paper import Paper
from Exceptions import OutOfInkError, PenCapacityError


class Pen:
    def __init__(self, ink_capacity: int = 4096) -> None:
        self._ink_amount = 0

        if ink_capacity <= 0:
            raise PenCapacityError('You entered incorrect pen capacity!')
        self._ink_capacity = ink_capacity

    @property
    def ink_amount(self) -> int:
        return self._ink_amount

    @property
    def ink_capacity(self) -> int:
        return self._ink_capacity

    def write(self, paper: 'Paper', msg: str) -> None:
        start = 0

        if self._ink_amount <= 0:
            raise OutOfInkError('Refill your pen!')

        new_msg = msg[start:self._ink_amount]
        paper.add_content(new_msg)
        self._ink_amount = self._ink_capacity - paper.symbols

    def refill(self) -> None:
        self._ink_amount = self._ink_capacity
